### PR TITLE
JAMES-3191 Key registration is slow

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by Arsnael** [comment]
Should we have some performance testing to backup this change?

**Comment by chibenwa** [comment]
> Should we have some performance testing to backup this change?

Probably.

**Comment by chibenwa** [comment]
I am trying to reproduce the initial cause of slowness, and that is non trivial.

**Comment by Arsnael** [comment]
test this please

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java [hunk 1]
```diff
@@ -125,20 +125,25 @@ void stop() {
...
+    Mono<Registration> register(MailboxListener.ReactiveMailboxListener listener, RegistrationKey key) {
         LocalListenerRegistry.LocalRegistration registration = localListenerRegistry.addListener(key, listener);
+
+        return registerIfNeeded(key, registration)
+            .thenReturn(new KeyRegistration(() -> {
+                if (registration.unregister().lastListenerRemoved()) {
+                    registrationBinder.unbind(key)
+                        .retryWhen(Retry.backoff(retryBackoff.getMaxRetries(), retryBackoff.getFirstBackoff()).jitter(retryBackoff.getJitterFactor()).scheduler(Schedulers.elastic()))
+                        .block();
+                }
```
Comment (blackheaven) [inline mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java:137]: A `block` in Reactor?

Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java:137]: Well that is in the unregister call back so it is not directly related to "register" but more "unregister"

I agree to push that work forward and reactify `unregister`.

(And if any :fr: people have some spare time, I would be happy to accept help on that)

**Comment by chibenwa** [comment]
```
[99324d8e5261182b21a7b102fb3a26b03beab3b4] [ERROR]   SerialTaskManagerWorkerTest.theWorkerShouldCancelAnInProgressTask:229 
[99324d8e5261182b21a7b102fb3a26b03beab3b4] No interactions wanted here:
[99324d8e5261182b21a7b102fb3a26b03beab3b4] -> at org.apache.james.task.SerialTaskManagerWorkerTest.theWorkerShouldCancelAnInProgressTask(SerialTaskManagerWorkerTest.java:229)
[99324d8e5261182b21a7b102fb3a26b03beab3b4] But found this interaction on mock 'listener':
[99324d8e5261182b21a7b102fb3a26b03beab3b4] -> at org.apache.james.task.SerialTaskManagerWorker.cancelled(SerialTaskManagerWorker.java:134)
[99324d8e5261182b21a7b102fb3a26b03beab3b4] ***
[99324d8e5261182b21a7b102fb3a26b03beab3b4] For your reference, here is the list of all invocations ([?] - means unverified).
[99324d8e5261182b21a7b102fb3a26b03beab3b4] 1. -> at org.apache.james.task.SerialTaskManagerWorker.run(SerialTaskManagerWorker.java:113)
[99324d8e5261182b21a7b102fb3a26b03beab3b4] 2. -> at org.apache.james.task.SerialTaskManagerWorker.handleExecutionError(SerialTaskManagerWorker.java:88)
[99324d8e5261182b21a7b102fb3a26b03beab3b4] 3. [?]-> at org.apache.james.task.SerialTaskManagerWorker.cancelled(SerialTaskManagerWorker.java:134)
```

Related to https://github.com/linagora/james-project/pull/3384

test this please

File: mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java (hunk 1 above)
Comment (chibenwa) [reply mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java]: While reactifying unregistration might be beneficial, I believe the true important thing here is a ` .subscribeOn(Schedulers.elastic())` call (that I contributed here). unregister is not called by reactor code thus the criticity is low and we might choose to delay this refactoring.

**Comment by chibenwa** [comment]
Yes. Due to unmanaged errirs resulting to a nack  we ended up with an infinite loop in mail processing.

Rabbitmq was the fist middleware to break  causing the mentionned slowness.

**Comment by chibenwa** [comment]
> @mbaechler added the requires-load-testing label 2 hours ago

Given the developments around this topic (that not the cause of RabbitMQ related performance issues), why do you think it requires load testing?

**Comment by mbaechler** [comment]
> > @mbaechler added the requires-load-testing label 2 hours ago
> 
> Given the developments around this topic (that not the cause of RabbitMQ related performance issues), why do you think it requires load testing?

Nope, go ahead and merge that

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**